scenario/state.py

```diff
@@ -140,6 +140,83 @@ def copy(self) -> "Self":
         return copy.deepcopy(self)
 
 
+@dataclasses.dataclass(frozen=True)
+class CloudCredential:
+    auth_type: str
+    """Authentication type."""
+
+    attributes: Dict[str, str] = dataclasses.field(default_factory=dict)
+    """A dictionary containing cloud credentials.
+
+    For example, for AWS, it contains `access-key` and `secret-key`;
+    for Azure, `application-id`, `application-password` and `subscription-id`
+    can be found here.
+    """
+
+    redacted: List[str] = dataclasses.field(default_factory=list)
+    """A list of redacted secrets."""
+
+    @classmethod
+    def from_dict(cls, d: Dict[str, Any]) -> "CloudCredential":
+        """Create a new CloudCredential object from a dictionary."""
```

Comment:
well, it isn't any dictionary, is it? I'd either make it a TypedDict or, perhaps better, unpack the dict in the function signature:

`def from_dict(cls, auth_type:str, attrs: Dict[...], redacted: List[str])`

This way users who already have the data structure can do `from_dict(**d)`, but you get a clear error if the dictionary structure is wrong.

Reply:
In ops, `credential_get` hook tool returns a json from which we create this object. I agree here we don't have to `from_dict`, but it's still nice to keep it if the user wants to simply copy the json and use that for test. User can create the CloudSpec object by instantiating the dataclass directly anyway, this isn't disallowed. Shall we keep it?